lib/srv/forward/sshserver.go

```diff
@@ -636,8 +637,34 @@ func (s *Server) newRemoteClient(ctx context.Context, systemLogin string) (*trac
 	if err != nil {
 		return nil, trace.Wrap(err)
 	}
-
 	return client, nil
+
+}
+
+// signersWithSHA1Fallback returns the signers provided by signersCb and appends
+// the same signers with forced SHA-1 signature to the end. This behavior is
+// required as older OpenSSH version <= 7.6 don't support SHA-2 signed certificates.
+func signersWithSHA1Fallback(signersCb func() ([]ssh.Signer, error)) func() ([]ssh.Signer, error) {
+	return func() ([]ssh.Signer, error) {
+		signers, err := signersCb()
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+		combinedSigners := make([]ssh.Signer, len(signers), 2*len(signers))
+		// Copy the original signers. All of them should be using SHA2-512 for signing.
+		// Order is important here as we want SHA2-signers to be used first.
+		copy(combinedSigners[:], signers)
+
+		for _, signer := range signers {
+			if s, ok := signer.(ssh.AlgorithmSigner); ok && s.PublicKey().Type() == ssh.CertAlgoRSAv01 {
+				// If certificate signer supports Sign(rand io.Reader, data []byte) method,
+				// add SHA-1 signer.
+				combinedSigners = append(combinedSigners, &sshutils.LegacySHA1Signer{Signer: s})
+			}
+		}
+
+		return combinedSigners, nil
```

Comment:
Does the `copy(combinedSigners[:], signers)` call  is required  ? Is still makes the shallow copy of Singer object. 
What would you you say about following simplification:
```suggestion

                var ss []ssh.Signer
		for _, signer := range signers {
			if s, ok := signer.(ssh.AlgorithmSigner); ok && s.PublicKey().Type() == ssh.CertAlgoRSAv01 {
				// If certificate signer supports Sign(rand io.Reader, data []byte) method,
				// add SHA-1 signer.
				ss = append(ss, &sshutils.LegacySHA1Signer{Signer: s})
			}
		}

		return apend(signers, ss...), nil
```